File: config-auth-core-basic/assets/auth-core-basic-template/src/application/usecase/create-user.usecase.ts
import { Result, UseCase } from "__SHARED_PACKAGE_NAME__";
import { UserExistsQuery } from "../provider";
import {
	Password,
	PasswordCryptoProvider,
	PasswordRepository,
} from "../../password";
import { User, UserErrors, UserRepository } from "../../user";

export interface CreateUserIn {
	name: string;
	email: string;
	password: string;
}

export class CreateUserUseCase implements UseCase<CreateUserIn, void> {
	constructor(
		private readonly userRepo: UserRepository,
		private readonly passRepo: PasswordRepository,
		private readonly userExistsQuery: UserExistsQuery,
		private readonly passwordCryptoProvider: PasswordCryptoProvider,
	) {}

	async execute(data: CreateUserIn): Promise<Result<void>> {
		const userExistsResult = await this.userExistsQuery.execute({
			email: data.email,
		});
		if (userExistsResult.isFailure) {
			return userExistsResult.withFail;
		}

		if (userExistsResult.instance) {
			return Result.fail(UserErrors.EMAIL_ALREADY_EXISTS);
		}

		const hashedPassword = await this.passwordCryptoProvider.hash(data.password);
		
		const passResult = Password.tryCreate({ content: hashedPassword });
		if (passResult.isFailure) {
			return passResult.withFail;
		}

		const userResult = User.tryCreate({
			name: data.name,
			email: data.email,
		});
		if (userResult.isFailure) {
			return userResult.withFail;
		}

		const userCreatedResult = await this.userRepo.create(userResult.instance);
		if (userCreatedResult.isFailure) {
			return userCreatedResult.withFail;
		}

		const createdUser = await this.userRepo.findByEmail(data.email);
		if (createdUser.isFailure) {
			return createdUser.withFail;
		}

		const passSaveResult = await this.passRepo.create(
			passResult.instance,
			createdUser.instance.id,
		);
		if (passSaveResult.isFailure) {
			return passSaveResult.withFail;
		}

		return Result.ok();
	}
}

File: config-auth-core-basic-kt/assets/auth-core-basic-template-kt/src/main/kotlin/com/example/auth/application/usecase/CreateUserUseCase.kt
package com.example.auth.application.usecase

import com.example.auth.application.query.UserExistsQuery
import com.example.auth.domain.entity.Password
import com.example.auth.domain.entity.User
import com.example.auth.domain.provider.PasswordCryptoProvider
import com.example.auth.domain.repository.PasswordRepository
import com.example.auth.domain.repository.UserRepository
import com.example.shared.application.UseCase
import com.example.shared.domain.result.DomainResult
import com.example.shared.infrastructure.TransactionManager

data class CreateUserInput(
    val name: String,
    val email: String,
    val password: String,
    val admin: Boolean = false,
    val avatarUrl: String? = null
)

class CreateUserUseCase(
    private val userRepository: UserRepository,
    private val passwordRepository: PasswordRepository,
    private val userExists: UserExistsQuery,
    private val crypto: PasswordCryptoProvider,
    private val transactionManager: TransactionManager
) : UseCase<CreateUserInput, User> {
    override suspend fun execute(data: CreateUserInput): DomainResult<User> {
        if (userExists.existsByEmail(data.email)) {
            return DomainResult.failure(listOf("USER_ALREADY_EXISTS"))
        }

        val userResult = User.tryCreate(
            name = data.name,
            email = data.email,
            admin = data.admin,
            avatarUrl = data.avatarUrl
        )
        if (userResult.isFailure) {
            return userResult
        }
        val user = userResult.value!!

        val hash = crypto.hash(data.password)
        val password = Password.create(userId = user.id.value, hash = hash)

        val results = transactionManager.runInTransaction {
            listOf(
                userRepository.create(user),
                passwordRepository.create(password)
            )
        }

        val errors = results.flatMap { it.errors }
        if (errors.isNotEmpty()) {
            return DomainResult.failure(errors)
        }

        return DomainResult.success(user)
    }
}